pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -118,24 +140,83 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
 	return nil
 }
 
-func validateContainer(container corev1.Container) *apis.FieldError {
+func validateVolume(volume corev1.Volume) *apis.FieldError {
+	var errs *apis.FieldError
+	if volume.Name == "" {
+		errs = apis.ErrMissingField("name")
+	} else if len(validation.IsDNS1123Label(volume.Name)) != 0 {
+		errs = apis.ErrInvalidValue(volume.Name, "name")
+	}
+
+	vs := volume.VolumeSource
+	switch {
+	case vs.Secret != nil, vs.ConfigMap != nil:
+		// These are fine.
+	default:
+		errs = errs.Also(apis.ErrMissingOneOf("secret", "configMap"))
+	}
+	return errs
+}
+
+func validateContainer(container corev1.Container, volumes sets.String) *apis.FieldError {
 	if equality.Semantic.DeepEqual(container, corev1.Container{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
+
+	// Check that volume mounts match names in "volumes", that "volumes" has 100%
+	// coverage, and the field restrictions.
+	seen := sets.NewString()
+	var errs *apis.FieldError
+	for i, vm := range container.VolumeMounts {
+		// This effectively checks that Name is non-empty because Volume name must be non-empty.
+		if !volumes.Has(vm.Name) {
+			errs = errs.Also((&apis.FieldError{
+				Message: "volumeMount has no matching volume",
+				Paths:   []string{"name"},
+			}).ViaFieldIndex("volumeMounts", i))
+		}
+		seen.Insert(vm.Name)
+
+		if vm.MountPath == "" {
+			errs = errs.Also(apis.ErrMissingField("mountPath").ViaFieldIndex("volumeMounts", i))
+		} else if reservedPaths.Has(filepath.Clean(vm.MountPath)) {
```

Comment:
Sure.  I thought we could volume mount a subdirectory under another volume (e.g. have both `/foo` and `/foo/bar` as volumes), but perhaps not.

Regardless, I'll go with this since I'd rather have overly strict validation and relax than the other way around.

Reply:
You're right, you could mount a subdir under a volume mount, but I suspect users might be surprised by the result.